### ISSUE TITLE
Some questions about tokenizer of BeiT v2

### ISSUE THREAD

**Issue by lizhiustc** [microsoft/unilm#936]
****A really nice work !****
 
But I have some questions about practical experience.

(1).**Does  a new  tokenizer based vqkd  can be trained on a small dataset, such as 30k images ?** (  The pretrained tokenizer in paper will encode every image into 256 codes, which are too long for me.  )  I have try to use CLIP-B/32 as teacher and the patch size of tokenizer is also 32( the new tokenizer will encode every image into 49 codes).  However, with origin config (12 encoder layers and 3 decoder layers) , most codes will not be used. So I only use 6 encoder layers, 1 decoder layer and  codebook_n_emd is  set to 1000. Still about  50% tokens will not be used. So another questions :

(2).**How to increase the usage of codebook? Contuning to reduce the number of layers  or dimension of  hidden states?**

(3).To take it a step further，**with few codes (49) of every image and without teacher, can this tokenizer be trained** ? Since the patch size of teacher limites the  code-numbers of image. Just make sure the tokenizer will be converaged and the quality of reconstructed images is not very important for me. If you don't try this situation, can you give me some **intuitive suggestions** ?

@pengzhiliang

**Comment by pengzhiliang** [comment]
Hello, thanks for the attention!

>(1).Does a new tokenizer based vqkd can be trained on a small dataset, such as 30k images ?

In fact, we have not trained vq-kd on a small dataset, so it's hard to determine the impact. But we have trained vq-kd on ImageNet-21k, and subsequently used it for MIM. Unfortunately, it shows the comparable performance with the counterpart trained on ImageNet-1k. Therefore, the training dataset for vq-kd may be not  a key factor for MIM.

>  ( The pretrained tokenizer in paper will encode every image into 256 codes, which are too long for me. ) I have try to use CLIP-B/32 as teacher and the patch size of tokenizer is also 32 (the new tokenizer will encode every image into 49 codes).

When the image resolution is 224, the length of discreate codes is 196, not 256 as you mentioned, but it doesn't matter. If the patch size is small (or the code length is long) for you, I think there may be two methods to fix: 
1. reduce the input image resolution and keep patch size unchanged. In this way, you do not need to train a new tokenizer. **This is strategy that we used when the resolution is not 224.**
2. chose a suitable CLIP model, which is actually what you do. CLIP-B/32 is a good candidate. But I sorry to hear that the codebook usage is low.

>(2)How to increase the usage of codebook? Contuning to reduce the number of layers or dimension of hidden states? ( with origin config (12 encoder layers and 3 decoder layers) , most codes will not be used. So I only use 6 encoder layers, 1 decoder layer and codebook_n_emd is set to 1000. Still about 50% tokens will not be used.)
I find that reducing the dimensions of hidden states will not help to increase the usage of codebook. And reducing the number of layers will help. I set the encoder layers to 3 and decoder layers to 1. Train the model with batch size 512 and 1k epochs. Codebook_n_emd is set to 3000 and train dataset have 30k images. Other hyper-parameters keep unchanged. The usage can reach 99%.

It sounds you have found a more suitable choice. But it's better to perform MIM phase to determine the final performance, beacuase tokenzier training is not the final phase.

> (3).To take it a step further，with few codes (49) of every image and without teacher, can this tokenizer be trained ? 

Without teacher? I am sorry that I am confused about it, can you describe it more clearly?

> (4)However, I find that the test_quant_loss will decreases first and then increases，is it normal ?

I checked the log:
![image](https://user-images.githubusercontent.com/26346329/205359692-3d4da637-5132-4d71-8f1a-a41e95b79e8c.png)
x-axis is epoch not the iteration here. And I do not observe the phenomenon you described. 
And my reconstruction loss is much smaller than yours:
![image](https://user-images.githubusercontent.com/26346329/205360378-9a269745-88cb-44ca-ae5c-1103f6a4e8ea.png)

> (5) And I am also confused that why we need more batch size and epochs on a small dataset to train a tokenzier with high codebook usage ? Maybe something important I missed ？

Maybe it is just for the converge.

If you still have queations, please let me know.

**Comment by lizhiustc** [comment]
Really helps a lot ! Thank you !

**Comment by Yi-Lynn** [comment]
Hi @lizhiustc ,

I’m working on training my own tokenizer using non-image data and have encountered a very similar train/validation curve to what you’ve shown in question (4)  
[.](https://github.com/microsoft/unilm/issues/936#issuecomment-1334731186) 

Specifically, my `quant_loss` decreases initially but then starts to increase, while the `unused_code` first increases and then drops, ultimately resulting in a very low codebook utilization rate.

Could you please share how you resolved this issue in the end? Thank you!

Below shows my results (pink and green curves correspond to two different hype-parameter settings.)

![Image](https://github.com/user-attachments/assets/d658c1a8-ef1c-4c0f-85b5-8be9681324f7)

![Image](https://github.com/user-attachments/assets/2f5567fc-a83e-4702-88ec-067f10b7d576)
